DQMOffline/RecoB/src/EffPurFromHistos2D.cc

```diff
@@ -0,0 +1,422 @@
+#include "DQMOffline/RecoB/interface/EffPurFromHistos2D.h"
+#include "DQMOffline/RecoB/interface/Tools.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "TStyle.h"
+#include "TCanvas.h"
+
+#include <iostream>
+#include <math.h>
+
+#include "DQMServices/Core/interface/DQMStore.h"
+#include "DQMServices/Core/interface/MonitorElement.h"
+#include "FWCore/ServiceRegistry/interface/Service.h" 
+
+using namespace std;
+using namespace RecoBTag;
+
+EffPurFromHistos2D::EffPurFromHistos2D ( const std::string & ext, TH2F * h_d, TH2F * h_u,
+                                     TH2F * h_s, TH2F * h_c, TH2F * h_b, TH2F * h_g, TH2F * h_ni,
+                                     TH2F * h_dus, TH2F * h_dusg, TH2F * h_pu,
+				     const std::string& label, const unsigned int& mc, 
+				     int nBinX, double startOX, double endOX) :
+        fromDiscriminatorDistr(false),
+        mcPlots_(mc), doCTagPlots_(false), label_(label),
+	histoExtension(ext), effVersusDiscr_d(h_d), effVersusDiscr_u(h_u),
+        effVersusDiscr_s(h_s), effVersusDiscr_c(h_c), effVersusDiscr_b(h_b),
+        effVersusDiscr_g(h_g), effVersusDiscr_ni(h_ni), effVersusDiscr_dus(h_dus),
+        effVersusDiscr_dusg(h_dusg), effVersusDiscr_pu(h_pu),
+	nBinOutputX(nBinX), startOutputX(startOX), endOutputX(endOX)
+{
+  // consistency check
+  check();
+}
+
+EffPurFromHistos2D::EffPurFromHistos2D (const FlavourHistograms2D<double, double> * dDiscriminatorFC, 
+				    const std::string& label, const unsigned int& mc, 
+                                    DQMStore::IBooker & ibook, 
+                                    int nBinX, double startOX, double endOX) :
+  fromDiscriminatorDistr(true), 
+  mcPlots_(mc), doCTagPlots_(false), label_(label),
+  nBinOutputX(nBinX), startOutputX(startOX), endOutputX(endOX)
+{
+  histoExtension = "_"+dDiscriminatorFC->baseNameTitle();
+
+  discrNoCutEffic = new FlavourHistograms2D<double, double> (
```

Comment:
@acaudron, looking at the code it seems correct not to call `delete` for `EffFlavVsBEff_*` since these objects are booked inside DQMStore and are therefore not owned by `EffPurFromHistos`. However, for all other cases where `new` is called inside `EffPurFromHistos` or `EffPurFromHistos2D`, I would expect a corresponding `delete` statement. I would suggest you simply define these objects as `std::unique_ptr` instead of plain pointers and then you won't have to worry about the `delete` statements.